pkg/providers/v1/topology.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/cloud-provider-aws/pkg/providers/v1/iface"
+	"k8s.io/klog/v2"
+)
+
+const instanceTopologyManagerCacheTimeout = 24 * time.Hour
+
+// stringKeyFunc is a string as cache key function
+func topStringKeyFunc(obj interface{}) (string, error) {
+	// Type should already be a string, so just return as is.
+	return obj.(string), nil
+}
+
+type instanceTopologyManager struct {
+	ec2                     iface.EC2
+	unsupportedInstanceType cache.Store
+	unsupportedRegion       cache.Store
+}
+
+func newInstanceTopologyManager(ec2 iface.EC2) *instanceTopologyManager {
+	return &instanceTopologyManager{
+		ec2: ec2,
+		// These should change very infrequently, if ever, so checking once a day sounds fair.
+		unsupportedRegion:       cache.NewTTLStore(topStringKeyFunc, instanceTopologyManagerCacheTimeout),
+		unsupportedInstanceType: cache.NewTTLStore(topStringKeyFunc, instanceTopologyManagerCacheTimeout),
+	}
+}
+
+func (t *instanceTopologyManager) getNodeTopology(instanceType string, region string, instanceID string) (*ec2.InstanceTopology, error) {
+	if t.mightSupportTopology(instanceType, region) {
+		request := &ec2.DescribeInstanceTopologyInput{InstanceIds: []*string{&instanceID}}
+		topologies, err := t.ec2.DescribeInstanceTopology(request)
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case "UnsupportedOperation":
+					klog.Infof("ec2:DescribeInstanceTopology is not available in %s.", region)
+					// If region is unsupported, track it to avoid making the call in the future.
+					t.addUnsupportedRegion(region)
+					return &ec2.InstanceTopology{}, nil
+				case "UnauthorizedOperation":
+					// gracefully handle the DecribeInstanceTopology access missing error
+					klog.Warningf("Not authorized to perform: ec2:DescribeInstanceTopology, permission missing")
+					return &ec2.InstanceTopology{}, nil
```

Comment:
Should return `nil` here IMO, we didn't get a topology result

Reply:
Ya, I realized that it was more than just the topologies that are a part of the request and that this returns the whole response, so I backtracked this to return `nil`.